http-verbs-common/src/main/scala/uk/gov/hmrc/play/http/BodyCaptor.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.play.http
+
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.stream.scaladsl.{Flow, Sink}
+import akka.stream.stage._
+import akka.util.ByteString
+import org.slf4j.LoggerFactory
+
+// based on play.filters.csrf.CSRFAction#BodyHandler
+
+private class BodyCaptorFlow(
+  loggingContext  : String,
+  maxBodyLength   : Int,
+  withCapturedBody: ByteString => Unit
+) extends GraphStage[FlowShape[ByteString, ByteString]] {
+  val in             = Inlet[ByteString]("BodyCaptorFlow.in")
+  val out            = Outlet[ByteString]("BodyCaptorFlow.out")
+  override val shape = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+      private var buffer     = ByteString.empty
+      private var bodyLength = 0
+
+      setHandlers(
+        in,
+        out,
+        new InHandler with OutHandler {
+          override def onPull(): Unit =
+            pull(in)
+
+          override def onPush(): Unit = {
+            val chunk = grab(in)
+            bodyLength += chunk.length
+            if (buffer.size < maxBodyLength)
+              buffer ++= chunk
+            push(out, chunk)
+          }
+
+          override def onUpstreamFinish(): Unit = {
+            withCapturedBody(BodyCaptor.bodyUpto(buffer, maxBodyLength, loggingContext, isStream = true))
+            if (isAvailable(out) && buffer == ByteString.empty)
```

Comment:
I can't see any reason - and testing (with empty streams) suggests it's not needed.
I will remove.

Reply:
This is actually a hack since in bootstrap it effectively calls `Flow[ByteString].via(BodyCaptor.flow(...)).toMat(Sink.head[ByteString])(Keep.right)`
The `Sink.head` will fail if there is no body. Note, that `Flow[ByteString].toMat(Sink.head[ByteString])(Keep.right)` would also fail (which doesn't go through the "patched" flow).
I will update bootstrap to use `Sink.headOption[ByteString]` and remove this hack.